sdk/core/azure-mgmt-core/azure/mgmt/core/policies/_base.py

```diff
@@ -0,0 +1,119 @@
+        http_request = request.http_request
+        response = self.next.send(request)
+        if response.http_response.status_code == 409:
+            rp_name = self._check_rp_not_registered_err(response)
+            if rp_name:
+                url_prefix = self._extract_subscription_url(http_request.url)
+                if not self._register_rp(request, url_prefix, rp_name):
+                    return response
+                # Change the 'x-ms-client-request-id' otherwise the Azure endpoint
+                # just returns the same 409 payload without looking at the actual query
+                if "x-ms-client-request-id" in http_request.headers:
+                    http_request.headers["x-ms-client-request-id"] = str(uuid.uuid4())
+                response = self.next.send(request)
+        return response
+
+    @staticmethod
+    def _check_rp_not_registered_err(response):
+        try:
+            response = json.loads(response.http_response.text())
+            if response["error"]["code"] == "MissingSubscriptionRegistration":
+                match = re.match(r".*'(.*)'", response["error"]["message"])
+                return match.group(1)
+        except Exception:  # pylint: disable=broad-except
+            pass
+        return None
+
+    @staticmethod
+    def _extract_subscription_url(url):
+        """Extract the first part of the URL, just after subscription:
+        https://management.azure.com/subscriptions/00000000-0000-0000-0000-000000000000/
+        """
+        match = re.match(r".*/subscriptions/[a-f0-9-]+/", url, re.IGNORECASE)
+        if not match:
+            raise ValueError("Unable to extract subscription ID from URL")
+        return match.group(0)
+
+    @staticmethod
+    def _build_next_request(initial_request, method, url):
+        request = HttpRequest(method, url)
+        context = PipelineContext(
+            initial_request.context.transport, **initial_request.context.options
+        )
+        return PipelineRequest(request, context)
+
+    def _register_rp(self, initial_request, url_prefix, rp_name):
+        """Synchronously register the RP is paremeter.
+
+        Return False if we have a reason to believe this didn't work
+        """
+        post_url = "{}providers/{}/register?api-version=2016-02-01".format(
+            url_prefix, rp_name
+        )
+        get_url = "{}providers/{}?api-version=2016-02-01".format(url_prefix, rp_name)
+        _LOGGER.warning(
+            "Resource provider '%s' used by this operation is not "
+            "registered. We are registering for you.",
+            rp_name,
+        )
+        post_response = self.next.send(
+            self._build_next_request(initial_request, "POST", post_url)
+        )
+        if post_response.http_response.status_code != 200:
+            _LOGGER.warning("Registration failed. Please register manually.")
+            return False
+
+        while True:
+            time.sleep(10)
```

Comment:
My concern is here.  We only return when receiving if rp_info["registrationState"] == "Registered". Or we will run for ever?

Reply:
I can't imagine that ARM would forever refuse to register without doing an error at some point. In theory, I agree with you, in practical term, I don't see how this could be possible.

Reply:
Because we use the exact string comparison, if they return the registrationState other than 'Registered' (maybe in different language for example?) Will that cause the loop to run for ever"

Reply:
Technically, I didn't try to set Accept-Language to try it :)
And if so, the timeout would not actually make it work, it will bypass the issue. That's a different problem. I don't think this field is localized, at least if I read the french documentation it's not:
https://docs.microsoft.com/fr-fr/azure/azure-resource-manager/management/resource-providers-and-types

Reply:
Yes. I found same results in Chinese. :)